kimchi/src/circuits/expr.rs

```diff
@@ -3149,3 +3149,246 @@ pub mod test {
+                }
+                _ => panic!("unsupported"),
+            };
+            new_exp
+        }
+        fn reduce_degree_to_2_rec(
+            self,
+            rec: &mut ExprRecorder<C, Col>,
+        ) -> Expr<C, ColumnExtended<Col>> {
+            let trivial_map = |c| ColumnExtended::Inner(c);
+            let degree = self.degree(4) / 4;
+            if degree <= 2 {
+                return self.map_column(&trivial_map);
+            }
+
+            type E<C, Col> = Expr<C, ColumnExtended<Col>>;
+            let e = match self {
+                e @ Expr::VanishesOnLast4Rows
+                | e @ Expr::UnnormalizedLagrangeBasis(_)
+                | e @ Expr::Constant(_)
+                | e @ Expr::Cell(_) => e.map_column(&trivial_map),
+
+                Expr::Double(e) => E::<C, Col>::Double(Box::new(e.reduce_degree_to_2_rec(rec))),
+                Expr::Square(e) => E::<C, Col>::Square(Box::new(e.reduce_degree_to_1(rec))),
+                Expr::BinOp(op, e1, e2) => match op {
+                    op @ Op2::Add | op @ Op2::Sub => {
+                        let e1 = Box::new(e1.reduce_degree_to_2_rec(rec));
+                        let e2 = Box::new(e2.reduce_degree_to_2_rec(rec));
+                        E::<C, Col>::BinOp(op, e1, e2)
+                    }
+                    Op2::Mul => {
+                        let e1 = Box::new(e1.reduce_degree_to_1(rec));
+                        let e2 = Box::new(e2.reduce_degree_to_1(rec));
+                        E::<C, Col>::BinOp(Op2::Mul, e1, e2)
+                    }
+                },
+                //only supporting 1..=8 for simplicity, but arbitrary exponent could be
+                //implemented, ideally an efficient fixed exponent algorithm
+                Expr::Pow(_, 0) => {
+                    panic!("no");
+                }
+                Expr::Pow(e, 1) => E::<C, Col>::Pow(Box::new(e.reduce_degree_to_2_rec(rec)), 1),
+                Expr::Pow(e, p) => {
+                    let new_exp = e.pow_to_mul(p);
+                    new_exp.reduce_degree_to_2_rec(rec)
+                }
+                Expr::Cache(id, e) => {
+                    let e = e.reduce_degree_to_2_rec(rec);
+                    E::<C, Col>::Cache(id, Box::new(e))
+                }
+                Expr::IfFeature(flag, e1, e2) => {
+                    let e1 = Box::new(e1.reduce_degree_to_2_rec(rec));
+                    let e2 = Box::new(e2.reduce_degree_to_2_rec(rec));
+                    E::<C, Col>::IfFeature(flag, e1, e2)
+                }
+            };
+            e
+        }
+
+        fn reduce_degree_to_1(
+            self,
+            rec: &mut ExprRecorder<C, Col>,
+        ) -> Expr<C, ColumnExtended<Col>> {
+            let trivial_map = |c| ColumnExtended::Inner(c);
+            let degree = self.degree(4) / 4;
+            if degree <= 1 {
+                return self.map_column(&trivial_map);
+            }
+            let extended = |id| {
```

Comment:
Is there a reason to define this as a lambda at this step? It doesn't seem to capture any external variables, so why not just inline it into the last row of `reduce_degree_to_1`?